modules/apps/frontend-data-set/frontend-data-set-views-web/src/main/resources/META-INF/resources/js/FDSEntries.tsx

```diff
@@ -187,14 +189,28 @@ const FDSEntries = ({
 		label: string;
 	};
 
-	const ProviderRenderer = ({itemData}: {itemData: FDSEntry}) => {
+	const ProviderRenderer = ({itemData}: {itemData: TFDSEntry}) => {
 		const headlessResource = headlessResourcesMapRef.current.get(
 			itemData.entityClassName
 		);
 
 		return `${headlessResource?.name} (${headlessResource?.bundleLabel} ${headlessResource?.version})`;
 	};
 
+	const ViewsCountRenderer = ({itemData}: {itemData: TFDSEntry}) => {
+		const count = itemData[OBJECT_RELATIONSHIP.FDS_ENTRY_FDS_VIEW].length;
+
+		return (
+			<span
+				className={classNames('count', {
+					'count-zero': !count,
+				})}
+			>
+				{count}
+			</span>
+		);
+	};
```

Comment:
One recommendation would be to either move this component outside of `FDSEntries` or use a `useMemo` to reduce how often we are "re-creating" this component.

Not a blocker, but I it's also easier to read when we don't create new named components inside of other component's lifecycle

Reply:
In general, I can think of two cases when moving the component to root is necessary:
1. When we want to export the component.
2. When the component has some inner state. This would lead to bugs.

Without any inner states, the component will always refresh with the parent component, so we wouldn't gain any performance by moving it outside. Although it's technically non-zero cost to assign a variable on each refresh, it is negligible.

By keeping it inside parent component, we reduce code, by not needing to forward params. This makes maintenance easier, because change of params needs to be done in a single place. This can get out of hand with time, as people like to mutate props, rename them, add defaults etc.

In this particular example, that main benefit does not apply. We're not using anything from parent context. In this case, it is only about what is more readable, a subjective call. It is slightly worse inside, because it has one indentation more. But, I still prefer it inside, because I like to collapse all components in a file, to find my bearing. We often have large files with multiple components, so placing every sub-component in root makes it harder to read. But yeah, this is all subjective. It's close, I wouldn't mind either.

Reply:
I'm leaning towards extracting it, but it can also be done if there's one more similar case in the same file. Then I'd for sure extract both of them.

Reply:
> I still prefer it inside, because I like to collapse all components in a file, to find my bearing. We often have large files with multiple components, so placing every sub-component in root makes it harder to read. But yeah, this is all subjective. It's close, I wouldn't mind either.

Yeah this might just come down to how each person reads through a file. Personally I find it much more difficult to read when the components aren't extracted to the root of the file.

IMO the two things that make readability harder for react is
- Indentations with JSX. The less amount of line breaks and multi-line components, the easier to read. So when we declare components in the body of an existing component, we already lose 2-3 tabs.
- Count the number of lines between props declaration and where the main `return` starts. As the number of lines increases, I think it is harder to read.


This isn't a deal breaker for me, so feel to merge without this change. However, I think if we continue this pattern of "monolithe" components that declare several other "inner components" things will be harder to debug later. I also don't often see this pattern in other projects, I typically see new components declared at the root of the file, but I can't prove this really, just from my own experience.

Reply:
Your component wont have the same issue, since it isn't an input, but [this is interesting](https://levelup.gitconnected.com/code-review-avoid-declaring-react-component-inside-parent-component-1768a645f523).

[I made a codesandbox to see it](https://codesandbox.io/s/throbbing-sky-5517de?file=/src/App.js)

Reply:
Another anecdote from another thread I found online discussing the same issue of creating a new component in an existing component.

> This is absolutely an antipattern. React will be unable to reconcile the rendering of these components, as they are treated as new components every time.
> 
> It's pretty much equivalent to creating a new class inside a function, only to instantiate it once and return it. There would be absolutely zero point in doing so.
> 
> If you try using hooks in those components, they will also not behave as expected. For example, a useState hook would always be populated with the initial value, rendering you unable to change its state.

Reply:
https://react.dev/learn/your-first-component#nesting-and-organizing-components

See "pitfalls"

Reply:
Yeah, that's basically what I mentioned above: it is necessary to place the component at the top level if it has an inner state. The one in this PR does not, so it's _technically_ ok to have it inside. But, I can see how this can be considered an anti-pattern. It can cause some headaches in the future if somebody extends inner component, or copies the code for own case and does not know about limitations.

I'll refactor it to be on top level.

Reply:
Yeah I was aware of the inner state part. I think I was just surprised that the new react docs make a more overarching claim by not only saying "components with inner state".

They say the following, which was surprising to me.
> Components can render other components, but you must never nest their definitions

> ... define every component at the top level:

>  When a child component needs some data from a parent, [pass it by props](https://react.dev/learn/passing-props-to-a-component) instead of nesting definitions.


Anyways, thanks for refactoring! Was helpful for me to look up myself as well